synapse/lib/filepath.py

```diff
@@ -0,0 +1,371 @@
+
+            zfd = self.zipfd.open(child_path.strip('/'), mode='r')
+            zbuf = zfd.read(read_chunk_sz)
+            while zbuf:
+                self.tempfd.write(zbuf)
+                zbuf = zfd.read(read_chunk_sz)
+
+            self.tempfd.flush()
+            child_path = self.tempfd.name
+            self.child = _get_path_class(child_path)(child_path, child_remainder, parent=self)
+
+        elif self.remainder:
+            raise s_exc.NoSuchPath(path=os.path.join(self.path, self.remainder))
+
+class FilePathZipFile(FilePathZipDir):
+    def _type(self):
+        return 'zip.inner.file'
+
+    def _open(self, mode='r'):
+        return self.zipfd.open(self.path.lstrip('/'), mode='r')
+
+class FilePathZip(FilePathZipDir):
+    def __init__(self, path, remainder, parent=None):
+        zipfd = zipfile.ZipFile(path)
+        self.zippath = path
+
+        super(FilePathZip, self).__init__(zipfd, '/', remainder, parent=parent)
+
+    def _type(self):
+        return 'zip.file'
+
+    def _open(self, mode='r'):
+        return open(self.zippath, mode=mode)
+
+def zip_enumerate(zipfd):
+    '''
+    Return the files and dirs in a zip file
+    '''
+    files = set()
+    dirs = set()
+
+    for path in zipfd.namelist():
+        
+        for spath in _get_subpaths(path):
+            try:
+                info = zipfd.getinfo(spath.strip('/'))
+                files.add(spath)
+            except KeyError as e:
+                dirs.add(spath)
+
+    return dirs, files
+
+def tar_enumerate(tarfd):
+    '''
+    Return the files and dirs in a tar archive
+    '''
+    files = set()
+    dirs = set()
+
+    for path in tarfd.getnames():
+        
+        for spath in _get_subpaths(path):
+            info = tarfd.getmember(spath.strip('/'))
+            if info.isdir():
+                dirs.add(spath)
+            if info.isfile():
+                files.add(spath)
+                
+    return dirs, files
+
+def _get_subpaths(path):
+    '''
+    Returns a list of subpaths in a path, one for each level in the path
+    '''
+    path_parts = path.strip('/').split('/')
+
+    paths = []
+    for i in range(len(path_parts)):
+
+        ipath = os.path.join('/', *path_parts[:i+1])
+        paths.append(ipath)
+
+    return paths
+
+def _enum_path(path):
+    fpobj = _parse_path(path)
+
+    if fpobj:
+        return fpobj._type()
+    return None
+
+def _parse_path(path):
+    '''
+    Internal function to parse the incoming path
+    '''
+
+    if not path:
+        return None
+
+    path = genpath(path)
+
+    path_parts = path.strip('/').split('/')
+
+    fpobj = None
+    next_path = os.path.join('/', path_parts[0])
+    remainder = '/'.join(path_parts[1:])
+
+    try:
+        cls = _get_path_class(next_path)
+        fpobj = cls(next_path, remainder)
+
+        while fpobj.getChild():
+            fpobj = fpobj.getChild()
+    except s_exc.NoSuchPath as e:
+        return None
+
+    return fpobj
+
+def _open(path, mode='r'):
```

Comment:
is this prefixed with _ to prevent collision with the builtin open() function?  If so, we should probably rename it rather than imply that it's "internal".  Maybe openfile() or openpath()